crates/pera-core/src/dwallet_mpc/encrypt_user_share.rs

```diff
@@ -0,0 +1,125 @@
+use group::GroupElement;
+use homomorphic_encryption::GroupsPublicParametersAccessors;
+use pera_types::base_types::PeraAddress;
+use pera_types::dwallet_mpc_error::{DwalletMPCError, DwalletMPCResult};
+use pera_types::messages_dwallet_mpc::{
+    StartEncryptedShareVerificationEvent, StartEncryptionKeyVerificationEvent,
+};
+use std::marker::PhantomData;
+use twopc_mpc::languages::class_groups::{
+    construct_encryption_of_discrete_log_public_parameters, EncryptionOfDiscreteLogProofWithoutCtx,
+};
+use twopc_mpc::secp256k1;
+use twopc_mpc::secp256k1::class_groups::AsyncProtocol;
+use twopc_mpc::secp256k1::SCALAR_LIMBS;
+
+type SecretShareEncryptionProof = EncryptionOfDiscreteLogProofWithoutCtx<
+    SCALAR_LIMBS,
+    SECP256K1_FUNDAMENTAL_DISCRIMINANT_LIMBS,
+    SECP256K1_NON_FUNDAMENTAL_DISCRIMINANT_LIMBS,
+    secp256k1::GroupElement,
+>;
+
+/// Verifies that the given encrypted secret share matches the encryption of the dWallet's
+/// secret share, validates the signature on the dWallet's public share,
+/// and ensures the signing public key matches the address that initiated this transaction.
+pub(crate) fn verify_encrypted_share(
+    verification_data: &StartEncryptedShareVerificationEvent,
+) -> DwalletMPCResult<()> {
+    verify_signatures(&verification_data)?;
+    chain_verify_secret_share_proof(
+        &verification_data.encrypted_secret_share_and_proof,
+        &verification_data.dwallet_output,
+        &verification_data.encryption_key,
+    )
+    .map_err(|_| DwalletMPCError::EncryptedUserShareVerificationFailed)
+}
+
+/// Verifies that the `verification_data`'s public key is matching the initiator Sui address.
+pub(crate) fn verify_encryption_key(
+    verification_data: &StartEncryptionKeyVerificationEvent,
+) -> DwalletMPCResult<()> {
+    let public_key =
+        <Ed25519PublicKey as ToFromBytes>::from_bytes(&verification_data.sender_sui_pubkey)
+            .map_err(|e| DwalletMPCError::EncryptedUserShareVerificationFailed)?;
+    let derived_sui_addr = PeraAddress::from(&public_key);
+    if derived_sui_addr != verification_data.initiator {
+        return Err(DwalletMPCError::EncryptedUserSharePublicKeyDoesNotMatchAddress);
+    }
+    Ok(())
+}
+
+/// Verify the signature on the public share of the DWallet,
+/// and that the public key that signed the public user share is matching the address that initiated this TX.
+fn verify_signatures(
+    verification_data: &StartEncryptedShareVerificationEvent,
+) -> DwalletMPCResult<()> {
+    let dkg_output: <AsyncProtocol as twopc_mpc::dkg::Protocol>::DecentralizedPartyDKGOutput =
+        bcs::from_bytes(&verification_data.dwallet_output)?;
+    let signature =
+        <Ed25519Signature as ToFromBytes>::from_bytes(&verification_data.signed_public_share)
+            .map_err(|e| DwalletMPCError::EncryptedUserShareVerificationFailed)?;
+    let public_key =
+        <Ed25519PublicKey as ToFromBytes>::from_bytes(&verification_data.encryptor_ed25519_pubkey)
```

Comment:
Why it is a different field from `verify_encryption_key()`? And why is the logic duplicated snd the function not called?